design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,287 @@
+itself. This concept is also not new in Crossplane, as [complex workloads]
+operate in this manner. `KubernetesApplication` instances are first scheduled to
+a Kubernetes cluster by the scheduler controller, which causes them to pass the
+predicates of the application controller. It uses the template(s) in the
+`KubernetesApplication` instance to create one or more
+`KubernetesApplicationResource` objects, which are then picked up by the
+resource scheduler. It is responsible for actually creating the
+Kubernetes-native (e.g. `Deployment`, `ConfigMap`, etc.) objects in the target
+cluster, then managing them as a managed reconciler does for an external
+resource. In this scenario, the Kubernetes-native objects are the external
+resources.
+
+## Goals
+
+* Create a common pattern that can be used across infrastructure stacks (and
+  potentially the existing complex workload controllers) that utilize the
+  Kubernetes API as the infrastructure provider API
+* Demonstrate the usage of this pattern with Rook
+
+## Differences from `KubernetesApplication`
+
+While the `KubernetesApplication` pattern detailed above provides a starting
+point for this design, a key difference is that `KubernetesApplication` does not
+adhere to the separation of concern structure of other Crossplane resources
+because Kubernetes resources are not modeled in a `claim <-> portable class <->
+non-portable class <-> managed <-> external` workflow. Kubernetes-native
+infrastructure resources, on the other hand, *should* be modeled with the same
+separation of concern that other cloud-provider resources expose.
+
+## Workflow Design
+
+The general design from a user experience perspective should differ very little
+from utilizing any other infrastructure provider stack. Developers should be
+able to request the creation of an abstract resource type via a claim, and
+cluster operators should be able to define the classes of service available to
+satisfy that claim kind.
+
+As an example, Rook provides the option to provision [Minio] object storage in a
+Kubernetes cluster. Assuming you have installed Rook and the corresponding Minio
+operator into a cluster, you should be able to dynamically provision a Minio
+object storage cluster into the Kubernetes cluster from your Crossplane control
+cluster. The experience would look as follows:
+
+1. Create a Kubernetes `Provider` for target cluster:
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: demo-provider-rook
+  namespace: rook-infra-dev
+type: Opaque
+data:
+  endpoint: MY_K8S_ENDPOINT
+  username: MY_K8S_USERNAME
+  password: MY_K8S_PASSWORD
+  clusterCA: MY_K8S_CLUSTER_CA
+  clientCert: MY_K8S_CLIENT_CERT
+  clientKey: MY_K8S_CLIENT_KEY
+  token: MY_K8S_TOKEN
+---
+apiVersion: kubernetes.crossplane.io/v1alpha2
+kind: Provider
+metadata:
+  name: demo-kubernetes
+  namespace: azure-infra-dev
+spec:
+  credentialsSecretRef:
+    name: demo-provider-rook
+```
```

Comment:
I wonder whether we should support a parameter like `thisCluster: true` on the Kubernetes Provider. This flag would instruct any connection logic to use [in cluster config](https://github.com/kubernetes/client-go/tree/95884bf844a94d573a28c0136282f4f8d0d4b8ae/examples/in-cluster-client-configuration), bypassing the need to read in a Secret.

Another (better?) approach could be to add a `credentialsServiceAccountRef` field to the spec, which would configure a (Kubernetes) service account whose bearer token should be used to provision managed resources in the local cluster, allowing you to use a different service account for each Kubernetes provider from the one that Crossplane is running as.

Another (maybe even better?) approach could be to just use `credentialsSecretRef` as we do today, but ensure our connection logic handles the case in which the supplied credentials secret is a Kubernetes [service account secret](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#service-account-tokens), not a Crossplane `KubernetesCluster` secret, i.e.:

```yaml
data:
  ca.crt: (APISERVER'S CA BASE64 ENCODED)
  namespace: ZGVmYXVsdA==
  token: (BEARER TOKEN BASE64 ENCODED)
```

This enables the following user flows.

Managing Rook resources in a Crossplane provisioned external cluster:
1. Provision a `KubernetesCluster` resource claim.
1. Create a Kubernetes `Provider` whose `credentialsSecretRef` references the `KubernetesCluster` `writeConnectionSecretToRef`.
1. Provision your Rook resource using the aforementioned `Provider`.

Managing Rook resources in an existing external cluster:
1. Create a Kubernetes `Provider` whose `credentialsSecretRef` references a hand curated `Secret` using the same format as that of a `KubernetesCluster` claim.
1.  Provision your Rook resource using the aforementioned `Provider`

Managing Rook resources in the Crossplane control cluster:
1. Create a Kubernetes `ServiceAccount` and grant it the RBAC permissions required to do Rook things.
1. Create a Kubernetes `Provider` whose `credentialsSecretRef` references the aforementioned `ServiceAccount` token.
1. Provision your Rook resource using the aforementioned `Provider`.

Reply:
@hasheddan If you're on board with the above (my preference is making the connection logic that parses a `credentialsSecretRef` also handle Service Account secrets) it would be good to capture that in this design.

Reply:
Really big fan of this design :) Will make sure to update the doc to reflect it.